src/components/implementation/no_interface/cFE_booter/ostask.c

```diff
@@ -0,0 +1,815 @@
+{
+	if (OS_TaskGetId() == main_delegate_thread_id) { return OS_ERR_INVALID_ID; }
+
+	struct cfe_task_info *task_info = &cfe_tasks[sl_thd_thdid(sl_thd_curr())];
+	task_info->delete_handler = function_pointer;
+
+	return OS_SUCCESS;
+}
+
+int32
+OS_TaskDelay(uint32 millisecond)
+{
+	cycles_t wakeup = sl_now() + sl_usec2cyc(millisecond * 1000);
+	sl_thd_block_timeout(0, wakeup);
+	return OS_SUCCESS;
+}
+
+int32
+OS_TaskSetPriority(uint32 task_id, uint32 new_priority)
+{
+	if (new_priority > 255 || new_priority < 1) { return OS_ERR_INVALID_PRIORITY; }
+
+	struct sl_thd *thd = sl_thd_lkup(task_id);
+	if (!thd) { return OS_ERR_INVALID_ID; }
+
+	union sched_param_union sp = {.c = {.type = SCHEDP_PRIO, .value = new_priority}};
+	sl_thd_param_set(thd, sp.v);
+
+	return OS_SUCCESS;
+}
+
+int32
+OS_TaskRegister(void)
+{
+	if (OS_TaskGetId() == main_delegate_thread_id) { return OS_ERR_INVALID_ID; }
+
+	return OS_SUCCESS;
+}
+
+
+int32
+OS_TaskGetIdByName(uint32 *task_id, const char *task_name)
+{
+	if (!task_id || !task_name) return OS_INVALID_POINTER;
+
+	thdid_t i;
+	for (i = 1; i < SL_MAX_NUM_THDS; i++) {
+		struct sl_thd *thd = sl_thd_lkup(i);
+		if (!thd || thd->state == SL_THD_FREE) continue;
+		if (strcmp(cfe_tasks[i].osal_task_prop.name, task_name) == 0) {
+			*task_id = i;
+			return OS_SUCCESS;
+		}
+	}
+
+	return OS_ERR_NAME_NOT_FOUND;
+}
+
+int32
+OS_TaskGetInfo(uint32 task_id, OS_task_prop_t *task_prop)
+{
+	if (!task_prop) { return OS_INVALID_POINTER; }
+
+	struct sl_thd *thd = sl_thd_lkup(task_id);
+
+	// TODO: Fix this ugly workaround
+	if (!thd || thd->state == SL_THD_FREE) { return OS_ERR_INVALID_ID; }
+
+	struct cfe_task_info *task_info = &cfe_tasks[task_id];
+	*task_prop = task_info->osal_task_prop;
+
+	return OS_SUCCESS;
+}
+
+/*
+** Main thread waiting API
+*/
+
+/*
+** OS-specific background thread implementation - waits forever for events to occur.
+**
+** This should be called from the BSP main routine / initial thread after all other
+** board / application initialization has taken place and all other tasks are running.
+*/
+void
+OS_IdleLoop(void)
+{
+	while (1) sl_thd_block(0);
+}
+
+/*
+** OS_ApplicationShutdown() provides a means for a user-created thread to request the orderly
+** shutdown of the whole system, such as part of a user-commanded reset command.
+** This is preferred over e.g. ApplicationExit() which exits immediately and does not
+** provide for any means to clean up first.
+*/
+void
+OS_ApplicationShutdown(uint8 flag)
+{
+	PANIC("Unimplemented method!"); // TODO: Implement me!
+}
+
+/*
+** Mutex API
+*/
+
+struct mutex {
+	int               used;
+	struct sl_lock    lock;
+	OS_mut_sem_prop_t prop;
+};
+
+struct sl_lock mutex_data_lock = SL_LOCK_STATIC_INIT();
+struct mutex   mutexes[OS_MAX_MUTEXES];
+
+
+int32
+OS_MutSemCreate(uint32 *sem_id, const char *sem_name, uint32 options)
+{
+	int32 result = OS_SUCCESS;
+
+	sl_lock_take(&mutex_data_lock);
+
+	if (sem_id == NULL || sem_name == NULL) {
```

Comment:
do this and the next error check *before* taking the lock.  Always want to minimize lock hold time, while, of course, maintaining correctness.